File: java/src/main/java/cn/example/slidingwindow/$76_MinimumWindowSubstring.java
package cn.example.slidingwindow;

import java.util.HashMap;
import java.util.Map;

public class $76_MinimumWindowSubstring {
    public String minWindow(String s, String t) {
        String res = null;
        int counter = 0, min_len = Integer.MAX_VALUE, start = 0, end = 0;
        Map<Character, Integer> hashmap = new HashMap<>();
        for(int i = 0; i < t.length(); i++){
            hashmap.put(t.charAt(i), 1);
        }

        while (end < s.length()){
            if(hashmap.getOrDefault(s.charAt(end), 0) > 0){
                counter --;  // 最长长度，所以是减一
            }
            hashmap.put(s.charAt(end), hashmap.getOrDefault(s.charAt(end), 0) - 1);
            end ++;
            while(counter == 0){
                if(end - start < min_len){
                    min_len = end -start;
                    res = s.substring(start, end);
                }
                if(hashmap.get(s.charAt(start)) > 0){
                    counter ++;
                }
                hashmap.put(s.charAt(start), hashmap.getOrDefault(s.charAt(start), 0) + 1);
                start ++;
            }
        }

        return res;
    }

}

File: python/slidingwindow/$76_MinimumWindowSubstring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        from collections import defaultdict
        hashmap = defaultdict(int)
        for i in t:
            hashmap[i] += 1
        start = 0
        end = 0
        min_len = float('inf')  # python中无穷大的表示
        counter = len(t)
        res = ""
        while end < len(s):
            if hashmap[s[end]] > 0:
                counter -= 1
            hashmap[s[end]] -= 1  # 这里用的是减，counter代表还未匹配到的t中字符串的数量
            end += 1
            while counter == 0:  # 满足条件时循环
                if end - start < min_len:  # 记录位置
                    min_len = end - start
                    res = s[start: end]
                if hashmap[s[start]] == 0:  # 此处只有等于零的才是t中元素，别的元素都小于零
                    counter += 1
                hashmap[s[start]] += 1
                start += 1
        return res


ans = Solution().minWindow("ADOBECODEBANC", "ABC")
